drivers/flash/soc_flash_numaker_rmc.c

```diff
@@ -0,0 +1,275 @@
+}
+
+/*
+ * Erase a flash memory area.
+ *
+ * param dev       Device struct
+ * param offset    The address's offset
+ * param len       The size of the buffer
+ * return 0       on success
+ * return -EINVAL erroneous code
+ */
+
+static int flash_numaker_erase(const struct device *dev, off_t offset, size_t len)
+{
+	struct flash_numaker_data *dev_data = dev->data;
+	uint32_t rc = 0;
+	unsigned int key;
+	int page_nums = (len / RMC_FLASH_PAGE_SIZE);
+	uint32_t addr = dev_data->flash_block_base + offset;
+
+	/* return SUCCESS for len == 0 (required by tests/drivers/flash) */
+	if (!len) {
+		return 0;
+	}
+
+	/* Validate range */
+	if (!flash_numaker_is_range_valid(offset, len)) {
+		return -EINVAL;
+	}
+
+	/* check alignment and erase only by pages */
+	if (((addr % RMC_FLASH_PAGE_SIZE) != 0) || ((len % RMC_FLASH_PAGE_SIZE) != 0)) {
+		return -EINVAL;
+	}
+
+	/* take semaphore */
+	if (k_sem_take(&dev_data->write_lock, K_NO_WAIT)) {
+		return -EACCES;
+	}
+
+	SYS_UnlockReg();
+	key = irq_lock();
+	while (page_nums) {
+		/* erase page */
+		if (RMC_Erase(addr)) {
+			LOG_ERR("Erase flash page failed or erase time-out");
+			rc = -EIO;
+			goto done;
+		}
+		page_nums--;
+		addr += RMC_FLASH_PAGE_SIZE;
+	}
+
+done:
+	SYS_LockReg();
+	irq_unlock(key);
+	/* release semaphore */
+	k_sem_give(&dev_data->write_lock);
+
+	return rc;
+}
+
+/*
+ * Read a flash memory area.
+ *
+ * param dev       Device struct
+ * param offset    The address's offset
+ * param data      The buffer to store or read the value
+ * param length    The size of the buffer
+ * return 0       on success,
+ * return -EIO     erroneous code
+ */
+static int flash_numaker_read(const struct device *dev, off_t offset, void *data, size_t len)
+{
+	struct flash_numaker_data *dev_data = dev->data;
+	uint32_t addr = dev_data->flash_block_base + offset;
+
+	/* return SUCCESS for len == 0 (required by tests/drivers/flash) */
+	if (!len) {
+		return 0;
+	}
+
+	/* Validate range */
+	if (!flash_numaker_is_range_valid(offset, len)) {
+		return -EINVAL;
+	}
+
+	/* read flash */
+	memcpy(data, (void *)addr, len);
+
+	return 0;
+}
+
+static int32_t flash_numaker_block_write(uint32_t u32_addr, uint8_t *pu8_data, int block_size)
+{
+	int32_t retval;
+	uint32_t *pu32_data = (uint32_t *)pu8_data;
+
+	SYS_UnlockReg();
+	if (block_size == 4) {
+		retval = RMC_Write(u32_addr, *pu32_data);
+	} else if (block_size == 8) {
+		retval = RMC_Write(u32_addr, *pu32_data) |
+			 RMC_Write(u32_addr + 4, *(pu32_data + 1));
+	} else {
+		retval = -1;
+	}
+	SYS_LockReg();
+
+	return retval;
+}
+
+static int flash_numaker_write(const struct device *dev, off_t offset, const void *data, size_t len)
+{
+	struct flash_numaker_data *dev_data = dev->data;
+	uint32_t rc = 0;
+	unsigned int key;
+	uint32_t addr = dev_data->flash_block_base + offset;
+	int block_size = flash_numaker_parameters.write_block_size;
+	int blocks = (len / flash_numaker_parameters.write_block_size);
+	uint8_t *pu8_data = (uint8_t *)data;
```

Comment:
do you need the casting? you are discarding the `const`

Reply:
Thanks of your advice, const is mandatory.